go/vt/vtgate/engine/coerce.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+	"fmt"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+)
+
+// Coerce is used to change types of incoming columns
+type Coerce struct {
+	Source Primitive
+	Types  []*evalengine.Type
+}
+
+var _ Primitive = (*Coerce)(nil)
+
+func (c *Coerce) RouteType() string {
+	return c.Source.RouteType()
+}
+
+func (c *Coerce) GetKeyspaceName() string {
+	return c.Source.GetKeyspaceName()
+}
+
+func (c *Coerce) GetTableName() string {
+	return c.Source.GetTableName()
+}
+
+func (c *Coerce) GetFields(
+	ctx context.Context,
+	vcursor VCursor,
+	bvars map[string]*querypb.BindVariable,
+) (*sqltypes.Result, error) {
+	res, err := c.Source.GetFields(ctx, vcursor, bvars)
+	if err != nil {
+		return nil, err
+	}
+	c.setFields(res)
+
+	return res, nil
+}
+
+func (c *Coerce) setFields(res *sqltypes.Result) {
+	for i, t := range c.Types {
+		if t == nil {
+			continue
+		}
+
+		typ := t.Type()
+		field := res.Fields[i]
+		_, flags := sqltypes.TypeToMySQL(typ)
+		if t.Nullable() {
+			flags |= int64(querypb.MySqlFlag_NOT_NULL_FLAG)
+		}
+
+		field.Type = typ
+		field.Flags = uint32(flags)
+		field.Charset = uint32(t.Collation())
+		field.ColumnLength = uint32(t.Size())
+		field.Decimals = uint32(t.Scale())
+	}
```

Comment:
Seems like you copied and pasted this from `evalengine`. Could you please make this a helper method there (that takes the field as an argument) as to not duplicate code?